### ISSUE TITLE
Issue with Hot reload

### ISSUE THREAD

**Issue by rkleivel** [flet-dev/flet#1556]
**Description**
First! Thanks so much for an AWESOME library :+1: 

I cannot seem to get hot reload work reliably in Ubuntu native window.

**Code example to reproduce the issue:**
Using the counter example from [flet docs](https://flet.dev/docs/):

```python
import flet as ft

def main(page: ft.Page):
    page.title = "Flet counter example"
    page.vertical_alignment = ft.MainAxisAlignment.CENTER

    txt_number = ft.TextField(value="0", text_align=ft.TextAlign.RIGHT, width=100)

    def minus_click(e):
        txt_number.value = str(int(txt_number.value) - 1)
        page.update()

    def plus_click(e):
        txt_number.value = str(int(txt_number.value) + 1)
        page.update()

    page.add(
        ft.Row(
            [
                ft.IconButton(ft.icons.REMOVE, on_click=minus_click),
                txt_number,
                ft.IconButton(ft.icons.ADD, on_click=plus_click),
            ],
            alignment=ft.MainAxisAlignment.CENTER,
        )
    )

ft.app(target=main)
```

**Describe the results you received:**
* When starting my app in OS native window with `flet run counter.py` the app stops responding if I change the code (for example just adding a blank line) and save (see screenshot below)
* If I start the app in browser with `flet run --web --port 1234 counter.py` the app reloads with new code after saving change as expected.

![image](https://github.com/flet-dev/flet/assets/15140187/9beee87e-93ea-4b8e-b0f0-b8617bc22bee)

Waiting long enough I sometimes get a response like this
```bash
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Stack Overflow
#0      String._concatAll (dart:core-patch/string_patch.dart)
#1      _StringBase._interpolate (dart:core-patch/string_patch.dart)
#2      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart)
#3      FletAppErrorsHandler.onError (package:flet/src/flet_app_errors_handler.dart)
#4      main.<anonymous closure> (package:flet_client/main.dart)
#5      FlutterError.reportError (package:flutter/src/foundation/assertions.dart)
#6      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart)
#7      FletAppErrorsHandler.onError (package:flet/src/flet_app_errors_handler.dart)
#8      main.<anonymous closure> (package:flet_client/main.dart)
#9      FlutterError.reportError (package:flutter/src/foundation/assertions.dart)
#10     ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart)
#11     FletAppErrorsHandler.onError (package:flet/src/flet_app_errors_handler.dart)
#12     main.<anonymous closure> (package:flet_client/main.dart)
.
.
.
#98216  main.<anonymous closure> (package:flet_client/main.dart)
#98217  FlutterError.reportError (package:flutter/src/foundation/assertions.dart)
#98218  ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart)
#98219  FletAppErrorsHandler.onError (package:flet/src/flet_app_errors_handler.dart)
#98220  main.<anonymous closure> (package:flet_client/main.dart)
#98221  FlutterError.reportError (package:flutter/src/foundation/assertions.dart)
#98222  ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart)
#98223  FletAppErrorsHandler.onError (package:flet/src/flet_app_errors_handler.dart)
#98224  main.<anonymous closure> (package:flet_client/main.dart)
#98225  PlatformDispatcher._dispatchError (dart:ui/platform_dispatcher.dart)
#98226  _onError (dart:ui/hooks.dart)

```

**Describe the results you expected:**
Would expect hot reload in OS native window to work like in browser

**Additional information you deem important (e.g. issue happens only occasionally):**
It seems to work as expected like 1 of 20 times.

**Flet version (`pip show flet`)**:

```
Name: flet
Version: 0.8.0
Summary: Flet for Python - easily build interactive multi-platform apps in Python
Home-page: 
Author: Appveyor Systems Inc.
Author-email: hello@flet.dev
License: Apache-2.0
Location: /home/roald/KMM/CODE_TESTING/flet-testing-py3_10/.venv/lib/python3.10/site-packages
Requires: copier, flet-runtime, packaging, pydantic, qrcode, watchdog, websocket-client, websockets
Required-by: 

```

**Operating system**:

Reproduced on 2 different computers:

1.
  * Ubuntu 20.04 python3.8.10

2.
  * Ubuntu 22.04 python3.10.6
  * Ubuntu 22.04 python3.11.4

**Additional environment details:**

**Comment by hexbenjamin** [comment]
glad to hear it's being looked at! you should know - the same issue is happening for me on win10

**Comment by FeodorFitsner** [comment]
I'm testing hot reload with the latest Flet on Ubuntu 22.04 and it's working quite well:


https://github.com/flet-dev/flet/assets/5041459/4fb9855c-3135-4780-8b9a-260518f9822d

**Comment by rkleivel** [comment]
Thanks for looking into this :) 
Here is a recording reproducing the problem:
[flet-run-ubuntu-22.04-hangs-on-hot-reload.webm](https://github.com/flet-dev/flet/assets/15140187/8667064d-4f8c-4c29-aef3-1ea2140d83a3)

NOTE: When the native window starts hanging, the flet process (I highlight it in the htop window in video) takes all cpu of one core. (In this recording the impression gets disturbed by also the screen recorder taking a lot of cpu, but when I'm not recording the highlighted flet process is the dominant cpu consumer)

I notice that your video filename indicates that you have an arm processor(?) I have Intel i7 - not sure if that is of any interest. 

Thankful for any ideas!

**Comment by rkleivel** [comment]
Trying to narrow down the root cause I have tried to reproduce the issue in 4 different environments:
* virtualbox, brand new Ubuntu 22.04 guest running on the Ubuntu 22.04 host where I recorded the video in previous comment
  * result: Issue Reproduced
* a windows 10 installation with Anaconda on the same computer (dual-boot) as in previous comment
  * result: Issue NOT reproduced
* a computer with xUbuntu 22.04
  * result: Issue Reproduced
* another smaller laptop with windows 10 / Anaconda
  * result: Issue NOT reproduced

**Comment by rkleivel** [comment]
Issue is now also reproduced on Raspberry Pi OS (Raspbian based on Debian 11 / Bullseye) on a Raspberry Pi 4 with ARM processor.

Before I potentially waste more of our valuable time on this I would highly appreciate if @FeodorFitsner or someone in the community would validate my reproduction procedure shown in the video below. Is this correct installation and use of flet? Or am I misunderstanding something basic here? 

Video shows installation of flet in a virtual environment on a fresh Ubuntu 22.04 installation, then create and run a basic test app. When saving an update to the app file, hot reload does not work, and flet starts to eat cpu and memory.

Thanks :)

https://github.com/flet-dev/flet/assets/15140187/0df3c94b-f4bf-4017-b9d7-4386959dddb2

**Comment by rkleivel** [comment]
> glad to hear it's being looked at! you should know - the same issue is happening for me on win10

@hexbenjamin - Would you be able to demonstrate (for example in a similar video) this issue on win10? I have tested on 2x win10 computers (with Anaconda) and not been able to reproduce there. 
Wish you a nice day :)

**Comment by sausix** [comment]
Assuming it's the same bug then it's NOT related to hot reload.
I've spent a few hours debugging flet and also _watchdog_ and it boils down to the handler function `restart_program` in _cli/commands/run.py_:

`self.p.send_signal(signal.SIGTERM)`

The SIGTERM does not work as expected and flutter's event loop gets broken or stuck somehow. Didn't dig deeper. Sorry.

And as I said it's not a hot reload problem because even without `-d` option you can trigger the bug by:

`kill -s SIGTERM pid` (as pid lookup the process id of `python -u [your-app.py]`

Can you confirm that @rkleivel ? Thank you.

Seem to affect UNIX systems: Linux and MacOS and it's probably a flutter bug.

Also a lot of bugreports seem to be a duplicate of this:
#1879
#1945 
#1975 
#2072 
#2111 
What about pinning this issue to reduce more bugreports?

**Comment by rlhk** [comment]
Hot reload never worked for me on macOS native.
For this reason I have to always use web mode or iOS Flet viewer app for dev.
Plus Safari client has serious memory leak for my OpenCV based video player component. So Chrome client becomes my dev default.

NOTE: no memory leak observed on macOS native.

**Comment by rkleivel** [comment]
> And as I said it's not a hot reload problem because even without `-d` option you can trigger the bug by:
> 
> `kill -s SIGTERM pid` (as pid lookup the process id of `python -u [your-app.py]`
> 
> Can you confirm that @rkleivel ? Thank you.

@sausix  - I can confirm that behaviour here also. But I do not have the knowledge needed to verify the conclusion you draw from it.

**Comment by sausix** [comment]
@rkleivel 
> But I do not have the knowledge needed to verify the conclusion you draw from it.

It's a problem not directly related to _hot reload_. And probably a bug in flutter. That's an important step and you helped to confirm that.

**Comment by ndonkoHenri** [comment]
Try this and let me know please:
https://ndonkohenri.medium.com/troubleshooting-hot-reload-in-the-latest-flet-versions-python-ae4eb3a6b0c6

**Comment by Klu1d** [comment]
> Try this and let me know please: https://ndonkohenri.medium.com/troubleshooting-hot-reload-in-the-latest-flet-versions-python-ae4eb3a6b0c6

I also have this issue with hot reloading in ubuntu 23.10. 

I followed the instructions in your article and my problem still remains

**Comment by rlhk** [comment]
> Try this and let me know please: https://ndonkohenri.medium.com/troubleshooting-hot-reload-in-the-latest-flet-versions-python-ae4eb3a6b0c6

Works flawless for me on macOS. Thanks for the solution!

**Comment by rkleivel** [comment]
> Try this and let me know please: https://ndonkohenri.medium.com/troubleshooting-hot-reload-in-the-latest-flet-versions-python-ae4eb3a6b0c6

Thanks, @ndonkoHenri  - I can confirm that this workaround works for me. Targeting the app with `ft.app(target=main, view=ft.AppView.WEB_BROWSER)` inside the script, **and** running with `flet run -p 1234 counter.py` solves the reload problem. Here is a video demo:


https://github.com/flet-dev/flet/assets/15140187/7f53f58b-c749-4b88-93c5-6b7662bbafa2

Another observation that could be worth mentioning:
The above demo is made on my HP laptop with i7 processor and an Nvidia graphics card.
I tested this also on another PC with **AMD Ryzen 9 on an ASUS motherbord with builtin graphics capabilities (no GPU), and was not able to reproduce the hot reload issue at all.** `ft.app(target=main)` together with `flet run counter.py` worked flawlessly there. I guess this indicates some platform relation(?)

**Comment by ndonkoHenri** [comment]
> I guess this indicates some platform relation(?)

Definitely, @rkleivel !

**Comment by Tatayoyoh** [comment]
Hi there ! 

I had the same problem (Linux Mint 21.1 Vera, Python 3.10.12, AMD Ryzen 7).

And it was resolved by adding a port `flet run chat.py -p 6000`

![Peek 17-12-2023 18-01](https://github.com/flet-dev/flet/assets/13559581/fcaf4c08-d82c-499d-a951-e710ee7ef99c)
